tensorflow/contrib/tensorrt/kernels/trt_calib_op.cc

```diff
@@ -0,0 +1,134 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/contrib/tensorrt/kernels/trt_calib_op.h"
+#include "tensorflow/contrib/tensorrt/resources/TRTInt8Calibrator.h"
+#include "tensorflow/contrib/tensorrt/resources/TRTResourceManager.h"
+#include "tensorflow/contrib/tensorrt/resources/TRTResources.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/framework/types.h"
+
+#if GOOGLE_CUDA
+#if GOOGLE_TENSORRT
+#include "cuda_runtime_api.h"
+#include "tensorrt/include/NvInfer.h"
+
+namespace tensorflow {
+namespace trt {
+TRTCalibOp::TRTCalibOp(OpKernelConstruction* context) : OpKernel(context) {
+  OP_REQUIRES_OK(context, context->GetAttr("segment_nodes", &segment_nodes_));
+  OP_REQUIRES_OK(context, context->GetAttr("input_names", &input_names_));
+  OP_REQUIRES_OK(context, context->GetAttr("resource_name", &resource_name_));
+};
+
+//  case statement for type
+#define TYPECASE(dt, X, Y)                                             \
+  case dt: {                                                           \
+    Y = (void*)X->flat<tensorflow::EnumToDataType<dt>::Type>().data(); \
+    break;                                                             \
+  }
+
+// macro to get tensor data address pointed by tensor_ptr into
+// void* dest_ptr
+#define GET_TENSOR_ADDRESS(tensor_ptr, dest_ptr)               \
+  {                                                            \
+    auto TENSOR_TYPE = tensor_ptr->dtype();                    \
+    switch (TENSOR_TYPE) {                                     \
+      TYPECASE(tensorflow::DT_FLOAT, tensor_ptr, dest_ptr);    \
+      TYPECASE(tensorflow::DT_HALF, tensor_ptr, dest_ptr);     \
+      TYPECASE(tensorflow::DT_INT8, tensor_ptr, dest_ptr);     \
+      default: {                                               \
+        LOG(FATAL) << "Unsupported Data type "                 \
+                   << tensorflow::DataTypeString(TENSOR_TYPE); \
+        break;                                                 \
+      }                                                        \
+    }                                                          \
+  }
+
+void TRTCalibOp::Compute(tensorflow::OpKernelContext* ctx) {
+  auto trt_rm = tensorflow::trt::TRTResourceManager::instance();
+  auto resmgr = trt_rm->getManager("TRTCalibOps");
```

Comment:
I think we don't need a special ResourceMgr here, instead we can can get one from ctx->resource_manager().

Reply:
This one is a singleton that is created during the conversion. does ctx->resource_manager() live outside session?

Reply:
Yes it does, it manages all the resource associated with the device during the lifetime of the device.

Reply:
Also, during the conversion we don't have access to context. Is there a way to get the same resource manager outside kernel execution?

Reply:
I will test this with our next PR. I am not sure if we have access to device at the conversion time. This manager is populated at the calibration conversion and consumed at the inference conversion. Please let it stay as it is right now. We will change as you propose if we can make it work with the conversion in the next PR.

Reply:
Ok, please add a TODO then. :)
A question regarding the conversion: why did you say during the conversion we don't have access to the context? Isn't the conversion done here inside OpKernel::Compute(ctx)?

Reply:
Conversion is done in ConvertGraphDefToTensorRT()